iio/iio_app/iio_app.c

```diff
@@ -413,33 +396,90 @@ int32_t iio_app_run_with_trigs(struct iio_ctx_attr *ctx_attrs,
-	iio_init_param.nb_ctx_attr = nb_ctx_attr;
-	status = iio_init(iio_desc, &iio_init_param);
+	iio_init_param.nb_devs = app->nb_devices;
+	iio_init_param.trigs = app->trigs;
+	iio_init_param.nb_trigs = app->nb_trigs;
+	iio_init_param.ctx_attrs = app->ctx_attrs;
+	iio_init_param.nb_ctx_attr = app->nb_ctx_attr;
+	status = iio_init(&(app->iio_desc), &iio_init_param);
 	if(status < 0)
 		goto error;
 
 	free(iio_init_devs);
 
-	do {
-		status = iio_step(*iio_desc);
-	} while (true);
+	return 0;
 error:
-	status = print_uart_error_message(&uart_desc, uart_init_par, status);
+	status = print_uart_error_message(&uart_desc, uart_init_par, app->user_uart_params, status);
 	return status;
 }
 
-int32_t iio_app_run(struct iio_ctx_attr *ctx_attrs, uint32_t nb_ctx_attr,
-		    struct iio_app_device *devices, uint32_t len)
+/**
+ * @brief Register devices for an iio application
+ *
+ * Configuration for communication is done in parameters.h
+ * @param app - the iio application descriptor
+ * @param app_init_param - the iio application initialization parameters
+ * @return 0 on success, negative value otherwise
+ */
+int32_t iio_app_init(struct iio_app_desc **app,
+	struct iio_app_init_param app_init_param)
 {
-	struct iio_desc	*iio_desc;
-	void *irq_desc = NULL;
+	struct iio_app_desc *application;
+	struct no_os_uart_init_param *luart_init_params;
+
+	application = (struct iio_app_desc *)calloc(1, sizeof(*application));
+	if (!application)
+		return -ENOMEM;
 
-	return iio_app_run_with_trigs(ctx_attrs, nb_ctx_attr, devices, len, NULL, 0,
-				      irq_desc, &iio_desc);
+	luart_init_params = (struct no_os_uart_init_param *)calloc(1, sizeof(*luart_init_params));
+	if (!luart_init_params) {
+		free(application);
+		return -ENOMEM;
+	}
+
+	application->ctx_attrs = app_init_param.ctx_attrs;
+	application->nb_ctx_attr = app_init_param.nb_ctx_attr;
+	application->devices = app_init_param.devices;
+	application->nb_devices = app_init_param.nb_devices;
+	luart_init_params->baud_rate = app_init_param.user_uart_init_params.baud_rate;
+	luart_init_params->size = app_init_param.user_uart_init_params.size;
+	luart_init_params->parity = app_init_param.user_uart_init_params.parity;
+	luart_init_params->stop = app_init_param.user_uart_init_params.stop;
+	application->user_uart_params = luart_init_params;
+	application->trigs = app_init_param.trigs;
+	application->nb_trigs = app_init_param.nb_trigs;
+	application->irq_desc = app_init_param.irq_desc;
+	application->app_callback = app_init_param.app_callback;
+	application->arg = app_init_param.arg;
+
+	*app = application;
+
+	return 0;
 }
 
+/**
+ * @brief Start an IIO application
+ *
+ * Configuration for communication is done through iio_app_init_param
+ * @param app - the iio application parameters
+ * @return 0 on success, negative value otherwise
+ */
+int32_t iio_app_run(struct iio_app_desc *app)
+{
+	int32_t	status;
+
+	status = iio_app_prepare(app);
+	if (status < 0)
+		return status;
+
+	do {
+		if (app->app_callback)
```

Comment:
I would rename app_callback to something more suggestive like `post_step` and move it such that it executes after iio_step.